go/vt/vttablet/tabletserver/throttle/throttler.go

```diff
@@ -0,0 +1,687 @@
+/*
+ Copyright 2017 GitHub Inc.
+
+ Licensed under MIT License. See https://github.com/github/freno/blob/master/LICENSE
+*/
+
+package throttle
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"vitess.io/vitess/go/vt/dbconnpool"
+	"vitess.io/vitess/go/vt/log"
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/connpool"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/tabletenv"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/throttle/base"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/throttle/config"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/throttle/mysql"
+
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	leaderCheckInterval         = 5 * time.Second
+	mysqlCollectInterval        = 100 * time.Millisecond
+	mysqlDormantCollectInterval = 5 * time.Second
+	mysqlRefreshInterval        = 10 * time.Second
+	mysqlAggreateInterval       = 100 * time.Millisecond
+
+	aggregatedMetricsExpiration   = 5 * time.Second
+	aggregatedMetricsCleanup      = 1 * time.Second
+	throttledAppsSnapshotInterval = 5 * time.Second
+	recentAppsExpiration          = time.Hour * 24
+
+	nonDeprioritizedAppMapExpiration = time.Second
+	nonDeprioritizedAppMapInterval   = 100 * time.Millisecond
+
+	dormantPeriod             = time.Minute
+	defaultThrottleTTLMinutes = 60
+	defaultThrottleRatio      = 1.0
+
+	maxPasswordLength = 32
```

Comment:
Based on experience, these are good numbers in production, and the risk for letting the user change these is that they'll use too small intervals (server too busy, and also meaningless since we have some error margin due to heartbeat injection), or too relaxed intervals (computation not reliable enough, causing high lags). I'm in the opinion of being opinionated to a reasonable degree, and feel comfortable with these values.